### ISSUE TITLE
no-useless-escape inside `new RegExp`

### ISSUE THREAD

**Issue by zhenyab** [eslint/eslint#6148]
**What version of ESLint are you using?**
v2.9.0
**What parser (default, Babel-ESLint, etc.) are you using?**
standard
**Please show your full configuration:**
```js
      "ecmaVersion": 6,
      "extends": "standard",
      "installedESLint": true,
      "plugins": [
        "standard"
      ],
      "env": {
        "node": true,
        "browser": true
      },
      "rules": {
        "comma-style": [
          "error"
          , "first"
        ]
        , "one-var": ["error", {
          "var": "always",
          "let": "always",
          "const": "always"
        }]
        , "space-before-function-paren": ["error",
          {"anonymous": "always", "named": "never"}]
      }
```
**What did you do? Please include the actual source code causing the issue.**
```js
`new RegExp('[^@]+@[^@]+\.[^@]+')`
```

**What did you expect to happen?**
Do not get any warnings or error.

**What actually happened? Please include the actual, raw output from ESLint.**
ESLint: Unnecessary escape character: `\.` (no-useless-escape)

**Comment by michaelficarra** [comment]
This is not a bug. You're probably looking to escape the `\`, not the `.`.

**Comment by zhenyab** [comment]
@michaelficarra This regex to check e-mail and I want really escape the `.` because I want to check that `.` (dot) is exists inside the string. If I use the alternate syntax of regex like so:
```
/[^@]+@[^@]+\.[^@]+/
```
then I do not get this warning.

**Comment by platinumazure** [comment]
@zhenyab The `new RegExp()` constructor takes a string literal, which has different rules for escaping than regular expression literals. So it makes sense that the rule warns for the string literal. As @michaelficarra noted, you need to escape the backslash in the string literal when you use the `new RegExp()` constructor.

**Comment by zhenyab** [comment]
@platinumazure Thanks for clarifications. Make sense.

**Comment by ilyavolodin** [comment]
@zhenyab Is this resolve then? Can the issue be closed?

**Comment by mysticatea** [comment]
Closing as this is not a bug. You should:

```js
new RegExp('[^@]+@[^@]+\\.[^@]+')
```

**Comment by greyvugrin** [comment]
I'm still seeing an issue when using @mysticatea 's example:
```
const sanitizePattern = new RegExp('[\[\]]', 'g')
```

^ both \ are now highlight above as unnecessary, with the /.../g syntax, only the first one was marked as an error.

**Comment by mysticatea** [comment]
@greyvugrin both `\[` and `\]` are unnecessary escape in string literals.

**Comment by greyvugrin** [comment]
@mysticatea in string literals, yes. But if the string is used as a regex the escape characters become necessary, otherwise the pattern is different.

 Without the escapes in `[\[\]]`, the pattern would be `[[]]`, which isn't the same.

**Comment by not-an-aardvark** [comment]
Your code is using `\[` and `\]` in a string literal, so the escapes are useless. Your string does not actually contain any backslash characters at the moment. The string that gets passed to the `RegExp` constructor is `[[]]`.

I assume you meant to use

```js
const sanitizePattern = new RegExp('[\\[\\]]', 'g')
```

**Comment by greyvugrin** [comment]
@not-an-aardvark thanks for the clarification, the double slash makes sense.

**Comment by ratracegrad** [comment]
I am using the RegExp to validate email addresses from http://emailregex.com/  which is 
/^(([^<>()\[\]\\.,;:\s@"]+(\.[^<>()\[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/
this generates two no-useless-escape errors. Any suggestion on how to overcome this without removing this warning in the eslintrc file?

**Comment by michaelficarra** [comment]
Remove the unnecessary backslashes before the dots. Inside a character class, they're just dots.

**Comment by ratracegrad** [comment]
@michaelficarra It is not completing about the blackslash in front of the periods but instead the first \ in the \[\] code as you see when you post it into http://eslint.org/demo/
If I remove the double blacksplashes in front of both periods it does not fix the error messages.
The only way to resolve is to remove the backslash and you end up with [\] which doesn't seem right. This code passes linting:
/^(([^<>()[\]\\.,;:\s@"]+(\.[^<>()[\]\\.,;:\s@"]+)*)|(".+"))@((\[[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}])|(([a-zA-Z\-0-9]+\.)+[a-zA-Z]{2,}))$/

**Comment by not-an-aardvark** [comment]
> The only way to resolve is to remove the backslash and you end up with [] which doesn't seem right. 

That is correct. In your original regex, `[` was uselessly escaped, because it's not necessary to escape a `[` character if it's already inside a code block.

As a sidenote: You can surround your code with ```` ``` ```` characters to make it easier to read:

````markdown
```
/your regex here/
```
````

This will ensure all the backslashes that you type get interpreted as intended, rather than being parsed as part of [markdown formatting](https://guides.github.com/features/mastering-markdown/). I've edited your two comments to add that.

**Comment by slaveofcode** [comment]
Have this problem too

    // somefile.js
    module.exports = (email) => {
        const re = /^(([^<>()\[\]\.,;:\s@\"]+(\.[^<>()\[\]\.,;:\s@\"]+)*)|(\".+\"))@(([^<>()[\]\.,;:\s@\"]+\.)+[^<>()[\]\.,;:\s@\"]{2,})$/i
        return re.test(email)
    }

I tried to move the regex pattern into `RegExp` object as a string argument and still got error, please help. Also is it possible to add some comment or something to ignore line from linting the code?

**Comment by michaelficarra** [comment]
@slaveofcode Don't escape `[` inside a character class or `"` anywhere. For ignoring lines, see the eslint documentation or just Google "eslint ignore line".
